server/mqtt.go

```diff
@@ -2322,7 +2310,15 @@ func (as *mqttAccountSessionManager) processSubs(sess *mqttSession, c *client,
 
 		// Note that if a subscription already exists on this subject, the
 		// existing sub is returned. Need to update the qos.
-		sub, err := lockAndInitializeSub(bsubject, bsid, f.qos, mqttDeliverMsgCbQoS0)
+		as.mu.Lock()
+		sess.mu.Lock()
+		sub, err := sess.processSub(c, bsubject, bsid,
+			isMQTTReservedSubscription(subject), f.qos, _EMPTY_, mqttDeliverMsgCbQoS0)
+		if err == nil {
+			as.serializeRetainedMsgsForSub(rms, sess, c, sub, trace)
```

Comment:
I think this is wrong, meaning that we used to do that for each sub (including shadow sub), here we are doing only for the main sub. Moreover, this needs to be done only when called with fromSubProto==true.